### PR TITLE
CMR-7235 APPEND_TO_FIELD operator for bulk granule update

### PR DESCRIPTION
**Pull request by jaybarra**
New APPEND_TO_FIELD operation

## S3Link
* if no S3 link data exists creates and adds, (identical to UPDATE_FIELD)
* if new urls are unique, will add new S3 links, preserving existing

### OPeNDAPLink
Because no more than 1 url of type cloud and 1 url of type on-prem may exist. behavior of APPEND_TO_FIELD is as follows
* if URL type is not present it will be added
* fails if existing type is present

### CODE REVIEW

File: ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj
Comment (mitchstartzel) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj:241]: I'm not sure I agree with this behavior. If the operations are identical, what's the point? It seems to add unneeded complexity. 

The use case for `append` in my eyes is "I want to add a link, but make sure I don't overwrite my existing links." For that reason, I think two alternatives are A) don't support append for OPeNDAP links, or B) have `append` fail if a link is added for an OPeNDAP link type which is already populated.

Comment (jaybarra) [reply ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj:241]: Could definitely do that. Need to have a discussion about the content of the ticket to finalize but making that change won't be difficult.

File: ingest-app/docs/api.md
Comment (mschmele) [inline ingest-app/docs/api.md:1116]:
```suggestion
If the URL passed to the update is already associated with the granule, the URL will not be duplicated.
```

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj
Comment (mschmele) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj:55]: Awkward phrasing here. Are you able to clarify this docstring?

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj
Comment (mschmele) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj:100]: No need for abbreviating this - it's more readable to use the full word

```suggestion
  [parsed online-accesses urls operation]
```

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj
Comment (mschmele) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj:133]:
```suggestion
  [gran-xml urls operation]
```

**Review by mschmele — COMMENTED** [review]
Logic looks good to me - I think it's worth finding a definite answer to Mitch's question before merging

File: ingest-app/docs/api.md
Comment (ygliuvt) [inline ingest-app/docs/api.md:1106]: If `APPEND_TO_FIELD` is identical to `UPDATE_FIELD` for OPeNDAPLink, maybe we shouldn't support it for OPeNDAP Links?

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/echo10.clj:148]: `Existing URLs will be preserved.` and `If an existing URL is included in the list of updates, it will be updated.` are contradicting statements.

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/umm_g.clj
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/umm_g.clj:33]: This will cause existing s3-url be duplicated if it is provided in the append list again.

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/umm_g.clj
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/s3/umm_g.clj:42]: change `update` to `append`.

File: ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj:281]: The only thing different in this function from the `:update_field:s3link` function is the call to `append-s3-url`. There is no need to copy paste all these code in this function. Refactor it to be a function and call it from both `update-granule-concept :append_to_field:s3link` and `update-granule-concept :update_field:s3link`.

File: system-int-test/test/cmr/system_int_test/ingest/granule_bulk_update/granule_bulk_update_test.clj
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/ingest/granule_bulk_update/granule_bulk_update_test.clj:470]: Add test that the existing url will not be duplicated (i.e. appear twice).

**Review by ygliuvt — CHANGES_REQUESTED** [review]
Need to address the review comments.

File: ingest-app/docs/api.md
Comment (jaybarra) [reply ingest-app/docs/api.md:1106]: That seems to be the consensus and logical outcome of the use case. I'll amend

File: ingest-app/docs/api.md
Comment (ygliuvt) [inline ingest-app/docs/api.md:1108]: This should be OPeNDAP urls. The same apply to the next line.

File: ingest-app/src/cmr/ingest/services/granule_bulk_update/opendap/echo10.clj
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/granule_bulk_update/opendap/echo10.clj:83]: put these validations in a function in `opendap-util`, so you can call it in here and in `umm-g` as well.

**Review by ygliuvt — APPROVED** [review]
Approved assuming the last minor refactoring comment is addressed.

**Review by mitchstartzel — CHANGES_REQUESTED** [review]
Needs a few docstring updates, and a documentation update to capture the functionality for appending pre-existing s3 links.

**Review by mitchstartzel — APPROVED** [review]

**Pull request merged**
